obal/tito_extensions/git_source_strategy.py

```diff
@@ -0,0 +1,183 @@
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+
+from datetime import datetime
+import glob
+import json
+import os
+import os.path
+import re
+import shutil
+import subprocess
+from zipfile import ZipFile
+
+from tito.builder.fetch import SourceStrategy
+from tito.common import error_out, debug, run_command
+
+class GitSourceStrategy(SourceStrategy):
+    """
+    Designed to be used for nightly or pull request builds only.
+    It first copies source files from git, then copies the source file(s)
+    over the top, so they're merged (patches etc can then be stored
+    in git).
+    """
+    def fetch(self):
+        if "source_dir" not in self.builder.args:
+            raise Exception("Specify '--arg source_dir=...'")
+
+        # Copy the live spec from our starting location. Unlike most builders,
+        # we are not using a copy from a past git commit.
+        self.spec_file = os.path.join(self.builder.rpmbuild_sourcedir,
+                    '%s.spec' % self.builder.project_name)
+        shutil.copyfile(
+            os.path.join(self.builder.start_dir, '%s.spec' %
+                self.builder.project_name),
+            self.spec_file)
+
+        gitrev = self._fetch_local()
+
+        for s in os.listdir(self.builder.start_dir):
+            if os.path.exists(os.path.join(self.builder.start_dir, s)):
+                shutil.copyfile(
+                    os.path.join(self.builder.start_dir, s),
+                    os.path.join(self.builder.rpmbuild_sourcedir, os.path.basename(s)))
+        print("  %s.spec" % self.builder.project_name)
+
+        replacements = []
+        src_files = run_command("find %s -type f" %
+              os.path.join(self.builder.rpmbuild_sourcedir, 'archive')).split("\n")
+
+        def filter_archives(path):
+            base_name = os.path.basename(path)
+            return ".tar" in base_name
+
+        for i, s in enumerate(filter(filter_archives, src_files)):
+            base_name = os.path.basename(s)
+            debug("Downloaded file %s" % base_name)
+
+            dest_filepath = os.path.join(self.builder.rpmbuild_sourcedir,
+                    base_name)
+            shutil.move(s, dest_filepath)
+            self.sources.append(dest_filepath)
+
+            # Add a line to replace in the spec for each source:
+            source_regex = re.compile("^(source%s:\s*)(.+)$" % i, re.IGNORECASE)
+            new_line = "Source%s: %s\n" % (i, base_name)
+            replacements.append((source_regex, new_line))
+
+        # Replace version in spec:
+        version_regex = re.compile("^(version:\s*)(.+)$", re.IGNORECASE)
+        self.version = self._get_version()
```

Comment:
Is this `self.version` used by anything else or could it be a local variable? There's a few other instances that I didn't check

Reply:
GitSourceStrategy inherits from SourceStrategy, which explicitly defines self.version.

this version variable is utilized by tito's mechanisms in various ways.

https://github.com/dgoodwin/tito/blob/master/src/tito/builder/fetch.py#L79